### PR TITLE
Added packet length transformers and changed Transformer API to use lists of ArrayBuffers.

### PR DESCRIPTION
**Pull request by blanu**
This is just a prototype. I have not written any unit tests. I have no checked for conformance to the style guide. I have not refactored the code to remove duplicated methods. I'm sending this patch in early to get some feedback and direction before I start cleaning up the code.

I added several transformers to act as functional tests. I have the following functionality in the transformers:
- add and remove fragmentation headers, without otherwise modifying the packet
- always increase the size, adding padding
- always decrease the size (causing fragmentation)
- normalize all packets to the same size
- randomize packet size within a specified range using a uniform distribution
- randomize packet size using a multinomial distribution

Additionally, the Transformer interface has been changed so that transform and restore now return ArrayBuffer[] instead of ArrayBuffer. This is to allow for fragmentation and defragmentation.

[Review on Reviewable](https://reviewable.io/reviews/uproxy/uproxy-lib/230)

### CODE REVIEW

File: src/churn-pipe/churn-pipe.ts [hunk 1]
```diff
@@ -60,6 +69,18 @@ var makeTransformer_ = (
      transformer = Fte.Transformer();
      } else */ if (name == 'caesar') {
```
Comment (mollyling) [inline src/churn-pipe/churn-pipe.ts]: I feel that we could turn this into a map from string to constructor.

File: src/churn/churn.ts [hunk 2]
```diff
@@ -271,9 +273,34 @@ export var filterCandidatesFromSdp = (sdp:string) : string => {
     private configurePipe_ = (key:number) : void => {
       this.pipe_ = freedom['churnPipe'](this.peerName);
       this.pipe_.on('mappingAdded', this.onMappingAdded_);
-      this.pipe_.setTransformer('caesar',
+      /*this.pipe_.setTransformer('none',
```
Comment (mollyling) [inline src/churn/churn.ts]: This is all a bit messy.  There seems to be quite a bit of magic going on in here.

File: src/churn/churn.ts [hunk 3]
```diff
@@ -288,6 +315,17 @@ export var filterCandidatesFromSdp = (sdp:string) : string => {
       this.havePipe_();
     }
 
+    private makeSampleTargetLengthDistribution_ = () : Array<number> => {
+      var result : Array<number> =[];
```
Comment (mollyling) [inline src/churn/churn.ts]: Style guide: types have colons right next to them.  So ":Array<number>" not ": Array<number>."  Not that you were claiming to adhere to the guide, but FYI.

File: src/fancy-transformers/defragmenter.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import logging = require('../logging/logging');
+import arraybuffers = require('../arraybuffers/arraybuffers');
+import Fragment = require('./fragment');
+
+var log :logging.Log = new logging.Log('fancy-transformers');
+
+class Defragmenter {
```
Comment (mollyling) [inline src/fancy-transformers/defragmenter.ts]: A class description would be nice.

File: src/fancy-transformers/defragmenter.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+  private tracker_ : {[index: string]: ArrayBuffer[]}={};
+  private counter_ : {[index: string]: number}={};
+  private complete_: string[]=[];
```
Comment (mollyling) [inline src/fancy-transformers/defragmenter.ts]: Again, style guide for spacing around colons.

File: src/churn/churn.ts [hunk 6]
```diff
@@ -288,6 +315,17 @@ export var filterCandidatesFromSdp = (sdp:string) : string => {
...
+      var prob = 1 / 1440;
+      var index=0;
```
Comment (mollyling) [inline src/churn/churn.ts]: There should be spaces around operators.

File: src/fancy-transformers/defragmenter.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+  public constructor() {}
+
+  public addFragment(fragment:Fragment) {
+    var hexid=arraybuffers.arrayBufferToHexString(fragment.id);
+    if(hexid in this.tracker_) { // A fragment for an existing packet
```
Comment (mollyling) [inline src/fancy-transformers/defragmenter.ts]: Vertical space is cheap, visual bandwidth is expensive.  Please put the comments on their own lines.

File: src/fancy-transformers/fragment.ts [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+}
+
+var split_ = (buffer:ArrayBuffer, firstLen:number) : Array<ArrayBuffer> => {
+  var bytes=new Uint8Array(buffer)
+  var lastLen : number = buffer.byteLength-firstLen;
+  var first = new Uint8Array(firstLen);
+  var last = new Uint8Array(lastLen);
+  var fromIndex : number = 0;
+  var toIndex : number = 0;
+  while(toIndex < first.length) {
```
Comment (mollyling) [inline src/fancy-transformers/fragment.ts]: Why isn't this a for() loop?

File: src/fancy-transformers/fragment.ts [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+  return [first.buffer, last.buffer];
+}
+
+/* Takes a two byte (network byte order) representation of a number and returns
+ * the number.
+ */
+ // TODO(bwiley): Byte order may be backward
+ // TODO(bwiley): Fix type error
+var decodeLength_ = (buffer:ArrayBuffer) : number => {
```
Comment (mollyling) [inline src/fancy-transformers/fragment.ts]: Perhaps generalize to ntohs/htons()?  `ntohs(buffer:ArrayBuffer, offset:int) :number` and `htons(value:number, buffer:ArrayBuffer, offset:int) :void`?

File: src/fancy-transformers/fragment.ts [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+      bytes[i] = Math.floor(Math.random()*255);
+    }
+    return bytes.buffer;
+  }
+
+  static decodeFragment = (buffer:ArrayBuffer) : Fragment => {
+//    log.debug('Decode fragment %1 %2', buffer.byteLength, length);
+    var parts = split_(buffer, 2);
+    var lengthBytes = parts[0];
+    var length = decodeLength_(lengthBytes);
```
Comment (mollyling) [inline src/fancy-transformers/fragment.ts]: These two variables: length and lengthBytes, don't tell me much about what they're the lengths of.

File: src/fancy-transformers/fragment.ts [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+      total=total+buffers[i].byteLength;
+    }
+
+    var result = new Uint8Array(total);
+    var toIndex=0;
+    for(var i=0; i<buffers.length; i++) {
+      var bytes=new Uint8Array(buffers[i]);
+      for(var fromIndex=0; fromIndex<buffers[i].byteLength; fromIndex++) {
+        result[toIndex]=bytes[fromIndex];
+        toIndex=toIndex+1;
```
Comment (mollyling) [inline src/fancy-transformers/fragment.ts]: you can do fromIndex++, toIndex++ in the last clause of the for-loop header.  Also another good place for a general copy routine.

File: src/fancy-transformers/defragmenter.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        fragments[fragment.index]=fragment.payload;
+        this.tracker_[hexid]=fragments;
+        this.counter_[hexid]=this.counter_[hexid]-1;
+        if(this.counter_[hexid]==0) {
+          this.complete_.push(hexid);
+        }
+      }
+    } else { // A fragment for a new packet
+      var fragments : ArrayBuffer[]=[];
+      for(var i=0; i<fragment.count; i++) {
```
Comment (mollyling) [inline src/fancy-transformers/defragmenter.ts]: I donno if you care about the difference but Array has a constructor that takes a single numerical arg, and results in an array of size `arg`, each initialized to `undefined`.

File: src/fancy-transformers/defragmenter.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        fragments.push(null);
+      }
+
+      fragments[fragment.index]=fragment.payload;
+      this.tracker_[hexid]=fragments;
+      this.counter_[hexid]=fragment.count-1;
+      if(this.counter_[hexid]==0) {
+        this.complete_.push(hexid);
```
Comment (mollyling) [inline src/fancy-transformers/defragmenter.ts]: Is this array intended to be grow-only, for the lifetime of the connection?

**Comment by blanu** [comment]
This pull request is out of date. I am closing it and creating an up-to-date one.

**Pull request closed without merging**